 
<div style="border:solid Chocolate 2px; padding: 40px">

Привет Амир, меня зовут Саша Атаков. Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем.
<br> Желательно реагировать на каждый мой комментарий ('исправил', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:

<div class="alert alert-block alert-success">
<b>Успех👍:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет🤔: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка❌:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>

-------------------

Будет очень хорошо, если ты будешь помечать свои действия следующим образом:
<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> ...
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Были внесены следующие изменения ...
</div>

<b><font color='orange'>Вот тебе полезная ссылка, вдруг подчерпнешь что-то новое для себя):</font></b>

* О логистической регрессии: https://habr.com/ru/company/io/blog/265007/
* Ну и случайный лес: https://alexanderdyakonov.wordpress.com/2016/11/14/%D1%81%D0%BB%D1%83%D1%87%D0%B0%D0%B9%D0%BD%D1%8B%D0%B9-%D0%BB%D0%B5%D1%81-random-forest/
    
<b><font color='orange'>Общее впечатление</font></b>
* Большое спасибо за проделанную работу. Видно, что ты отлично освоил тему.
* Отлично, что ноутбук хорошо структурирован. Приятно проверять такие работы.
* В работе я оставил несколько советов. Буду рад, если ты учтешь их.
* В конце прописал свой комментарий по всему проекту, ознакомься с ним, пожалуйста)
* С радостью отвечу на твои вопросы, касательно выданной тебе темы, если они есть. Жду новую версию проекта.)

<div class="alert alert-block alert-warning">
Привет! Саша, спасибо тебе за ревью. Надеюсь, что моя тетрадка действительно неплоха и у меня есть шансы после окончания курса устроиться на работу))))
</div>

<div class="alert alert-block alert-success">
<b>Успех👍 V2:</b> 
    
Привет). Твой проект приятен, честно). Работу найти будет сложновато, но у тебя есть хорошие шансы
</div>

# Рекомендация тарифов

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

## Откройте и изучите файл

In [1]:
# загружаем необходимые библиотеки
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score 


<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Хороший шаг - импортировать все необходимые библиотеки перед работой с проектом, ведь так ты точно не запутаешься в том, какие библиотеки ты уже импортировал, а какие стоит добавить. Молодец!
</div>

In [2]:
# загружаем датасет
df = pd.read_csv('/datasets/users_behavior.csv')

df.head()

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

In [3]:
# изучим датафрейм
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


Датасет загружен, в данных нет пропусков, тип данных соответсвует описываемому параметру.

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Замечательно, данные на базе и уже проверены!
</div>

## Разбейте данные на выборки

In [4]:
# выделим целевой показатель и параметры

features = df.drop('is_ultra', axis=1)
target = df['is_ultra']

In [5]:
# разобьем данные на 3 выборки: обучающую, валидационную и тестовую (60%, 20%, 20%).

features_train, features_test, target_train, target_test = train_test_split(features, target, test_size=0.25, random_state=123)
features_train, features_valid, target_train, target_valid = train_test_split(features_train, target_train, test_size=0.25, random_state=123)

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Вот это мне нравится, разделение на три выборки, молодец!
</div>

## Исследуйте модели

### Модель "дерево решений"

In [6]:
best_tree_model = None
best_accuracy_tree = 0
best_max_depth_tree = 1
best_min_sample_split_tree = 2
for depth in range(1,15):
    for sample_split in range(2,5):
        model_tree = DecisionTreeClassifier(max_depth=depth, min_samples_split=sample_split, random_state=123)
        model_tree.fit(features_train, target_train)
        predictions_valid_tree = model_tree.predict(features_valid)
        accuracy_tree = accuracy_score(target_valid, predictions_valid_tree)
        if best_accuracy_tree < accuracy_tree:
            best_accuracy_tree = accuracy_tree
            best_tree_model = model_tree
            best_max_depth_tree = depth
            best_min_sample_split_tree = sample_split
print (f'Лучшая модель "дерево решений" имеет accuracy = {best_accuracy_tree:.4f}, max_depth = {best_max_depth_tree}, min_sample_split = {best_min_sample_split_tree}.')

Лучшая модель "дерево решений" имеет accuracy = 0.7778, max_depth = 10, min_sample_split = 2.


<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Да, дерево обучено верно, так держать!
</div>

### Модель "случайный лес"

In [7]:
best_forest_model = None
best_accuracy_forest = 0
best_max_depth_forest = 1
best_min_sample_split_forest = 2
best_n_estimators = 1
for depth in range(1, 10):
    for min_sample in range(2,5):
        for est in range(5,50,5):
            model_forest = RandomForestClassifier(max_depth=depth, 
                                                  min_samples_split=min_sample, 
                                                  n_estimators=est, 
                                                  random_state=123)
            model_forest.fit(features_train, target_train)
            predictions_valid_forest = model_forest.predict(features_valid)
            accuracy_forest = accuracy_score(target_valid, predictions_valid_forest)
            if best_accuracy_forest < accuracy_forest:
                best_accuracy_forest = accuracy_forest
                best_forest_model = model_forest
                best_max_depth_forest = depth
                best_min_sample_split_forest = sample_split
                best_n_estimators = est
print (f'Лучшая модель "случайный лес" имеет accuracy = {best_accuracy_forest:.4f}, max_depth = {best_max_depth_forest}, min_sample_split = {best_min_sample_split_forest}, n_estimators = {best_n_estimators}.')
            



Лучшая модель "случайный лес" имеет accuracy = 0.7960, max_depth = 9, min_sample_split = 4, n_estimators = 15.


<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
И случайный лес обучен хорошо, замечательно)
</div>

<div class="alert alert-block alert-info">
<b>Совет🤔:</b> 

По дефолту у случайного леса 100 деревьев, а ты начал от 5 до 50. Может лучше пойти от 100 до, скажем, 500, с шагом 50
</div>

<div class="alert alert-block alert-warning">
Попробую твой совет в следующей ячейке
</div>

best_forest_model = None
best_accuracy_forest = 0
best_max_depth_forest = 1
best_min_sample_split_forest = 2
best_n_estimators = 1
for depth in range(1, 10):
    for min_sample in range(2,5):
        for est in range(100,500,50):
            model_forest = RandomForestClassifier(max_depth=depth, 
                                                  min_samples_split=min_sample, 
                                                  n_estimators=est, 
                                                  random_state=123)
            model_forest.fit(features_train, target_train)
            predictions_valid_forest = model_forest.predict(features_valid)
            accuracy_forest = accuracy_score(target_valid, predictions_valid_forest)
            if best_accuracy_forest < accuracy_forest:
                best_accuracy_forest = accuracy_forest
                best_forest_model = model_forest
                best_max_depth_forest = depth
                best_min_sample_split_forest = sample_split
                best_n_estimators = est
print (f'Лучшая модель "случайный лес" имеет accuracy = {best_accuracy_forest:.4f}, max_depth = {best_max_depth_forest}, min_sample_split = {best_min_sample_split_forest}, n_estimators = {best_n_estimators}.')

<div class="alert alert-block alert-warning">
Для ускорения твоей работы ячейку выше сделал текстовой. Вот такой результат я получил при выполнении кода:   

Лучшая модель "случайный лес" имеет accuracy = 0.7944, max_depth = 8, min_sample_split = 4, n_estimators = 150.   

<b>Очень долго, а качество модели лучше не стало.</b>
</div>

<div class="alert alert-block alert-success">
<b>Успех👍 V2:</b> 
    
Круто, что проверил мой совет). Жалко, что увеличении количества деревьев не помогло в улучшении модели. Но проверка того стоила) 
</div>

### Модель логистической регрессии

In [8]:
best_reg_model = None
best_accuracy_reg = 0
best_solver = None
best_max_iter = 0
for m_iter in range(100,1100,100):
    model_reg = LogisticRegression(random_state=123, solver='lbfgs', max_iter=m_iter)
    model_reg.fit(features_train, target_train)
    predictions_valid_reg = model_reg.predict(features_valid)
    accuracy_reg = accuracy_score(target_valid, predictions_valid_reg)
    if accuracy_reg > best_accuracy_reg:
        best_accuracy_reg = accuracy_reg
        best_reg_model = model_reg
        best_max_iter = m_iter
print (f'Лучшая модель логистической регрессии имеет accuracy = {best_accuracy_reg:.4f}, max_iter = {best_max_iter}.')

Лучшая модель логистической регрессии имеет accuracy = 0.7247, max_iter = 100.


<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Ну и последнаяя модель - логистическая регрессия, справилась с заданием)
</div>

<div class="alert alert-block alert-info">
<b>Совет🤔:</b> 

К логистической регрессии также проверяют параметр силы регуляризации (`C`). Можешь ее также проверить, ради интереса
</div>

<div class="alert alert-block alert-warning">
Сейчас проверю данный гиперпараметр
</div>

In [17]:
best_reg_model = None
best_accuracy_reg = 0
best_solver = None
best_max_iter = 0
best_C = 0
for m_iter in range(100,1100,100):
    for c in [1, 10, 100, 1000]:
        model_reg = LogisticRegression(random_state=123, solver='lbfgs', max_iter=m_iter, C=c)
        model_reg.fit(features_train, target_train)
        predictions_valid_reg = model_reg.predict(features_valid)
        accuracy_reg = accuracy_score(target_valid, predictions_valid_reg)
        if accuracy_reg > best_accuracy_reg:
            best_accuracy_reg = accuracy_reg
            best_reg_model = model_reg
            best_max_iter = m_iter
            best_C = c
print (f'Лучшая модель логистической регрессии имеет accuracy = {best_accuracy_reg:.4f}, max_iter = {best_max_iter}, C = {best_C}.')

Лучшая модель логистической регрессии имеет accuracy = 0.7247, max_iter = 100, C = 1.


<div class="alert alert-block alert-warning">
Результат остался прежним. Я так понимаю, что чем меньше С, тем лучше.
</div>

<div class="alert alert-block alert-success">
<b>Успех👍 V2:</b> 
    
Ну хотя бы не упал). А так, качество зависит не от того, меньше ли С. Чем больше параметр С, тем более сложные зависимости в данных может восстанавливать модель (интуитивно С соответствует "сложности" модели)
</div>

<div class="alert alert-block alert-info">
<b>Совет🤔 V2:</b> 

На самом деле было бы точнее брать значения от -100 до 1000 (лучше брать именно списком - [-100, -10, -1, 0, 1, 10 ...])
</div>

<div class="alert alert-block alert-warning">
<b> V2 </b>   
    
Саш, попробовал -100, -10, 0. Пшите ошибку, что отрицательным быть не может и на 0 делить нельзя. И флоат не получалось, хотя в доке к библиотеке там можно указывать числа с запятой, как я понял
</div>

# Выбор лучшей модели   

В ходе исследования мы построили 3 модели для задачи классификации, где основным параметром качества модели являлось значение **accuracy**.
Лучшая `модель "дерево решений"` имеет **accuracy = 0.7778**, max_depth = 10, min_sample_split = 2.   
Лучшая `модель "случайный лес"` имеет **accuracy = 0.7960**, max_depth = 9, min_sample_split = 4, n_estimators = 15.   
Лучшая `модель логистической регрессии` имеет **accuracy = 0.7247**, max_iter = 100.   

**Таким образом, при построении и обучении моделей на данном датасете, лучшей моделью стала** `модель "случайный лес"`**.**

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Значит все-таки лес - хорошо)
</div>

## Проверьте модель на тестовой выборке

In [10]:
# проверяем модель"случайный лес" на тестовой выборке
predictions_test_forest = best_forest_model.predict(features_test)
accuracy_forest = accuracy_score(target_test, predictions_test_forest)
print (f'Модель "случайный лес" на тестовых данных имеет accuracy = {accuracy_forest:.4f}')

Модель "случайный лес" на тестовых данных имеет accuracy = 0.8035


In [11]:
# для интереса оценим как поведут себя оставшиеся модели на тестовой выборке
# модель "дерево решений"
predictions_test_tree = best_tree_model.predict(features_test)
accuracy_tree = accuracy_score(target_test, predictions_test_tree)
print (f'Модель "дерево решений" на тестовых данных имеет accuracy = {accuracy_tree:.4f}')

Модель "дерево решений" на тестовых данных имеет accuracy = 0.7923


In [12]:
# модель логистической регрессии
predictions_test_reg = best_reg_model.predict(features_test)
accuracy_reg = accuracy_score(target_test, predictions_test_reg)
print (f'Модель логистической регрессии на тестовых данных имеет accuracy = {accuracy_reg:.4f}')

Модель логистической регрессии на тестовых данных имеет accuracy = 0.7450


**Лучшая обученная модель ("случайный лес") на тестовых данных также показала лучшие результаты! Accuracy данной модели на тестовой выборке = 0.8035.**

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Замечательно, мы проверили модели на тестовой выборке  и это круто, но однако...
</div>

<div class="alert alert-block alert-info">
<b>Совет🤔:</b> 

Если тебе надо проверить модель на тестовых значениях, то ты проверяешь только лучшую модель из обученных и проверенных на `valid` выборке. На тест попадает лучшая из лучших
</div>

<div class="alert alert-block alert-warning">
На будущее учту, спасибо)
</div>

<div class="alert alert-block alert-success">
<b>Успех👍 V2:</b> 
    
Да, это будет относительно полезно (наверное)
</div>

## (бонус) Проверьте модели на адекватность

Для проверки адекватности модели классификации необходимо сравнить ее со случайной моделью.

In [13]:
# импортируем DummyClassifier из Sklern, чтобы создать случайную модель
from sklearn.dummy import DummyClassifier

model_random = DummyClassifier(strategy='uniform', random_state=123)
model_random.fit(features_train, target_train)
predictions_random_val = model_random.predict(features_valid)
predictions_random_test = model_random.predict(features_test)

accuracy_random_val = accuracy_score(target_valid, predictions_random_val)
accuracy_random_test = accuracy_score(target_test, predictions_random_test)
print (f'Случайная модель на валидационных данных имеет accuracy = {accuracy_random_val:.4f}')
print (f'Случайная модель на тестовых данных имеет accuracy = {accuracy_random_test:.4f}')

Случайная модель на валидационных данных имеет accuracy = 0.4826
Случайная модель на тестовых данных имеет accuracy = 0.5050


**Accuracy** `случайной модели` на тестовых данных равна **0.5050**. Это значительно меньше выбранной лучшей модели (`"случайный лес"`- **accuracy = 0.8035**). Однако необходимо проверить, есть ли между данными значениями статистически значимые различия.   
Нам необходимо сравнить 2 выборки **predictions_random_test** и **predictions_test_forest** с категориальным типом данных. Для этого применяют хи-квадрат Пирсона. Рассчитаем р-value!


<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Да, это действительно бонус и приятный). И то, что ты собрался проверить на статическую разность - молодец)
</div>

In [14]:
from scipy.stats import chi2_contingency
import numpy as np
data = np.array([[predictions_random_test.sum(), (len(predictions_random_test) - predictions_random_test.sum())],
                [predictions_test_forest.sum(), (len(predictions_test_forest) - predictions_test_forest.sum())]])
statistics = chi2_contingency(data)
if statistics[1] > 0.05:
    print(f'Выбранная модель не адекватна, статистически значимых различий со случайной выборкой нет, p-value = {statistics[1]:.3f}')
else:
    print(f'Выбранная модель адекватна, статистически значимые различия со случайной выборкой выявлены, p-value = {statistics[1]:.3f}')


Выбранная модель адекватна, статистически значимые различия со случайной выборкой выявлены, p-value = 0.000


# Выводы
В ходе исследования мы построили 3 модели для задачи классификации (распределение пользователей по типам тарифов), где основным параметром качества модели являлось значение **accuracy**.   

`Модель "дерево решений"` имеет **accuracy = 0.7778**, max_depth = 10, min_sample_split = 2.   
`Модель "случайный лес"` имеет **accuracy = 0.7960**, max_depth = 9, min_sample_split = 4, n_estimators = 15.   
`Модель логистической регрессии` имеет **accuracy = 0.7247**, max_iter = 100.   

Таким образом, при построении и обучении моделей на данном датасете, лучшей моделью стала `модель "случайный лес"`.   

Лучшая обученная модель (`"случайный лес"`) на тестовых данных также показала лучшие результаты! Accuracy данной модели на тестовой выборке = 0.8035.   

При оценке адекватности разработанной модели проведено ее сранение со случайной моделью. Наша модель показала статистически значимо более высокие результаты качества, р-value = 0.000.

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Да, верно, твой вывод охватывает весь проект и это мне нравится, снимаю шляпу. Все четко и структурированно
</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [x] Весь код исполняется без ошибок
- [x] Ячейки с кодом расположены в порядке исполнения
- [x] Выполнено задание 1: данные загружены и изучены
- [x] Выполнено задание 2: данные разбиты на три выборки
- [x] Выполнено задание 3: проведено исследование моделей
    - [x] Рассмотрено больше одной модели
    - [x] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [x] Написаны выводы по результатам исследования
- [x] Выполнено задание 3: Проведено тестирование
- [x] Удалось достичь accuracy не меньше 0.75


<div style="border:solid Chocolate 2px; padding: 40px">
    
<h2><b>Общий вывод по проекту</b></h2>
    
Ты очень постарался, спасибо тебе за такую работу!)

В ходе проекта были выполнены все этапы проекта, главная цель достигнута. Конечно присутствует ряд моментов, которые необходимо рассмотреть, но ничего страшного, для это я здесь). Давай пройдемся по всем этапам твоего проекта чтобы понять, как ты справился:
    
<b>`Загрузка данных`</b>. С этим этапом никаких вопросов нет, здесь все сделано правильно: библиотеки загружены в отдельной ячейке над всем проектом, создан датасет, был выполнен базовый просмотр данных.
    
<b>`Исследуйте модели`</b>. На данном этапе ты реализовал три алгоритма МО:
    
- Дерево - самый базовый алгоритм, но иногда дающий приятные результаты. В нем ты попробовал гиперпараметры такие как: минимальное деление выборок и глубина. Довольно хорошие гиперпараметры, особенно глубина, ведь так мы проверяем большее количество факторов при определении ответа для дерева  
    
- Случайный лес - самый популярный алгоритм за счет своей точности, ведь решение принимает несколько деревьев. Ты проверил те же признаки, что и у дерева + количество деревьев. Да, количество деревьев хорошо улучшает точность прогноза, но и повышает время обучения и принятия решения. В твоем проекте стоит попробовать большее количество деревьев, ведь от 5 до 50 это ниже дефолтного состояния леса
    
- Логистическая регрессия - самый математичный алгоритм, который хорошо показывает обучение и предсказание моделей МО со стороны математики. Ты проверил количество итераций, хороший выбор). Но чаще проверяют именно силу регуляризации. Посмотри что это такое (долгая тема, предупреждаю сразу)
    
<b>`Проверка моделей`</b>. Тут ты проверил модели на тестовой выборке, но правильнее было бы проверить только лучшую модель, а не все сразу, ведь предыдущий этап позволил тебе найти эту лучшую модель). После этого была проверка на адекватность и тут все замечательно, молодец)
    
Вот так вот. Работа хорошая, могу принять при следующей твоей отправке (сейчас можешь просто ознакомиться с моими рекомендациями, а можешь что-то дополнить). Удачи!)

<div class="alert alert-block alert-warning">
Саша, еще раз спасибо за ревью! Можно твои советы:   
    
- какие модели в настоящее время пользуются наибольшей популярностью, с чего стоит начинать? Или брать как можно больше алгоритмов моделей и искать лучший?     
    
- среди данных моделей какие гиперпараметры еще смотрят или основные охвачены?   
    
- данное задание выполнено ниже среднего, средне или выше среднего? Хочется для себя объективно понять уровень на данном этапе обучения.
</div>

<div class="alert alert-block alert-info">
<b>Совет🤔 V2:</b> 

Ух, люблю вопросы, давай пройдемся по ним:
    
* Наиболее популярные модели, наверное линейная и логистическая регрессии, градиентный бустинг, нейронные сети (точно знаю, что BERT крайне эффективная), kmeans. На самом деле каждая модель подходит под определенный вид задач, так что нельзя сказать, что одна модель подходит под все задачи)
    
* 1) Для дерева и леса стоит еще помнить об атрибуте разбиения (criterion), так как он осуществляет важную функцию - при обучении множества в узле, данный атрибут следит за тем, чтобы в этих множествах было меньше дисбаланса классов значений, к которым мы и предсказываем что-либо. 2) Для логистической регрессии можно вспомнить о штрафе за ошибку для модели (penalty), так мы сможем дать модели пинка, если она будет давать не те ответы, которые мы бы хотели видеть и модель будет стараться дать более точный ответ). Также есть баланс классов - class_weight. Понадобится для того, чтобы помочь модели понять, что у нас один класс, больше другого и модель меняет веса для модели
    
* Ну оценивать работу я думаю не стоит, ведь я могу повести тебя не в ту сторону, какую тебе стоит идти. Поэтому я скажу тебе то, что ты крайне заинтересованный студент, коллега если тебе так удобнее), а это очень важно в нашей сфере, так ты сможешь искать что-то интересное и узнавать что-то новое. Также ты ценишь советы, а это уже позволяет тебе улучшать свои знания). Если ты не потеряешь этого же напора, то у тебя получится стать классным спецом в МО, удачи!
</div>

<div class="alert alert-block alert-warning">
Еще раз спасибо за софтскилы)    

Вообще я врач, надеюсь мои навыки в этом направлении помогут мне быть чуть лучше в датасайнс в медицине.
</div>